### PR TITLE
XW-904 | Replace $.ajax with ember-ajax

### PR DESCRIPTION
**Pull request by hakubo**
## Links

* https://wikia-inc.atlassian.net/browse/XW-904
* https://github.com/ember-cli/ember-ajax

## Description

This PR prepares us to use Ember Fastboot, but before we get there we need to have a way to call APIs with the same code both on client and server side. Ember ajax has a basic support for that and they are committed to improve that part of the lib. 

## Reviewers

@Wikia/x-wing

### CODE REVIEW

File: front/main/app/initializers/register-model.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import CategoryModel from '../models/wiki/category';
+import ArticleAddPhotoModel from '../models/article-add-photo';
+import ArticleEditModel from '../models/article-edit';
+import ArticleCommentsModel from '../models/article-comments';
+import WikiaInYourLangModel from '../models/wikia-in-your-lang';
+import UserModel from '../models/user';
+import RecentWikiActivityModel from '../models/recent-wiki-activity';
+import InfoboxBuilderModel from '../models/infobox-builder';
+import ImageReviewItemModel from '../models/image-review-item';
+import FandomModel from '../models/image-review';
```
Comment (rogatty) [inline front/main/app/initializers/register-model.js]: Something not right here :see_no_evil:

Comment (rogatty) [inline front/main/app/initializers/register-model.js]: I don't see `getModel(this, 'image-review')` anywhere. Is it TODO?

File: front/main/app/initializers/register-model.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import CuratedContentEditorModel from '../models/curated-content-editor';
+import CuratedContentEditorItemModel from '../models/curated-content-editor-item';
+
+export function initialize(application) {
+	application.register('model:article', ArticleModel);
+	application.register('model:category', CategoryModel);
+	application.register('model:article-add-photo', ArticleAddPhotoModel);
+	application.register('model:article-edit', ArticleEditModel);
+	application.register('model:article-comments', ArticleCommentsModel);
+	application.register('model:wikia-in-your-lang', WikiaInYourLangModel);
```
Comment (rogatty) [inline front/main/app/initializers/register-model.js]: Keep the alphabetical order, will be easier to manage.

File: front/main/app/initializers/register-model.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	application.register('model:infobox-builder', InfoboxBuilderModel);
+	application.register('model:image-review-item', ImageReviewItemModel);
+	application.register('model:fandom', FandomModel);
+	application.register('model:curated-content', CuratedContentModel);
+	application.register('model:curated-content-editor', CuratedContentEditorModel);
+	application.register('model:curated-content-editor-item', CuratedContentEditorItemModel);
+}
+
+export default {
+	name: 'register-model',
```
Comment (rogatty) [inline front/main/app/initializers/register-model.js]: `register-models`

File: front/main/app/models/curated-content-editor.js [hunk 4]
```diff
@@ -35,70 +36,51 @@ import CuratedContentEditorItemModel from '../models/curated-content-editor-item
...
 const CuratedContentEditorModel = Ember.Object.extend({
-	featured: null,
-	curated: null,
-	communityData: null,
-	optional: null,
-	isDirty: false
-});
+	ajax: Ember.inject.service(),
+	init() {
+		this._super(...arguments);
```
Comment (bognix) [inline front/main/app/models/curated-content-editor.js]: are we moving the initiation of all properties to init? or only the `non-primitive` ones?
It differs between models (this one has all and curated-content has some)

Comment (hakubo) [inline front/main/app/models/curated-content-editor.js]: I think we should only move `non-primitives` to init

File: front/main/app/models/search-images.js [hunk 5]
```diff
@@ -37,6 +37,7 @@ import Thumbnailer from 'common/modules/thumbnailer';
  */
 
 export default Ember.Object.extend({
+	ajax: Ember.inject.service(),
```
Comment (bognix) [inline front/main/app/models/search-images.js]: this one is still in progress?

File: front/main/app/models/wiki/article.js [hunk 6]
```diff
@@ -17,6 +18,7 @@ import {normalizeToWhitespace} from 'common/utils/string';
  */
 
 const ArticleModel = Ember.Object.extend({
+	ajax: Ember.inject.service(),
 	content: null,
 	basePath: null,
 	categories: [],
```
Comment (bognix) [inline front/main/app/models/wiki/article.js]: should it be in init?

File: front/main/app/models/wikia-in-your-lang.js [hunk 7]
```diff
@@ -3,55 +3,46 @@ import LanguagesMixin from '../mixins/languages';
...
-				localStorageConnector.setItem(
-					WikiaInYourLangModel.getCacheKey(browserLang),
-					JSON.stringify({
-						model: modelInstance,
-						timestamp: new Date().getTime()
-					})
-				);
+			// write to cache
+			localStorageConnector.setItem(
+				this.getCacheKey(browserLang),
```
Comment (bognix) [inline front/main/app/models/wikia-in-your-lang.js]: this method can be defined outside model class

**Comment by bkoval** [comment]
Is this touching the whole platform with one refactor?

Is it working well with promises rejection and error substates?

Will you update discussions as well? We use one abstraction layer on all our ajax calls (utils/ajax-call.js) so it seems like it should be fairly straightforward :)

**Comment by hakubo** [comment]
Thanks Bart for chiming in.

Yes - my intention was to use ember-ajax everywhere - that is something we have to do before fastboot. 

It works well with promises as this returns an RSVP promise.

I intend to write a longer summary of these changes - with longer explanation of the changes.
Yes, I want to refactor discussions also - did not have time last week to finish that

File: front/main/app/models/article-add-photo.js [hunk 8]
```diff
@@ -2,222 +2,198 @@ import Ember from 'ember';
...
-							photoImage: oFREvent.target.result,
-							photoName,
-							photoExtension
-						})
-					);
-				};
-				oFReader.onerror = function () {
-					reject();
-				};
+				return this;
```
Comment (rogatty) [inline front/main/app/models/article-add-photo.js]: Shouldn't this be `resolve(this)`?

File: front/main/app/models/article-comments.js [hunk 9]
```diff
@@ -12,15 +13,10 @@ export default Ember.Object.extend({
...
-			return new Ember.RSVP.Promise((resolve, reject) => {
-				Ember.$.ajax({
-					url: this.url(articleId, page),
-					success: (data) => {
-						this.setProperties(data.payload);
-						resolve(this);
-					},
-					error: (data) => reject(data)
-				});
+			return this.get('ajax').request(url(articleId, page)).then((data) => {
```
Comment (rogatty) [inline front/main/app/models/article-comments.js]: `this.url(`

File: front/main/app/models/curated-content-editor.js (hunk 4 above)
Comment (rogatty) [inline front/main/app/models/curated-content-editor.js]: So... Do you plan to move these back?

File: front/main/app/models/discussion/base.js [hunk 10]
```diff
@@ -90,24 +90,20 @@ export default Ember.Object.extend({
...
-				} else {
-					track(trackActions.UpvotePost);
-				}
-			},
-			error: () => {
-				entity.set('userData.hasUpvoted', hasUpvoted);
-			},
-			complete: () => {
-				this.upvotingInProgress[entityId] = false;
+		this.get('ajax')[method](
```
Comment (rogatty) [inline front/main/app/models/discussion/base.js]: There are only `post`, `put`, `patch` and `del` methods [available in ember-ajax](https://github.com/ember-cli/ember-ajax#http-verbed-methods) and we have:
```
method = hasUpvoted ? 'delete' : 'post'
```

File: front/main/app/models/discussion/forum.js [hunk 11]
```diff
@@ -22,25 +22,23 @@ const DiscussionForumModel = DiscussionBaseModel.extend(
...
-					this.handleLoadMoreError(err);
 				}
+			}).then((data) => {
+				this.get('data.entities').pushObjects(
+					Ember.get(data, '_embedded.doc:threads').map(
+						(newThread) => DiscussionPost.createFromThreadData(newThread)
+					)
+				);
+			}).catch((err) => {
+				this.handleLoadMoreError(err);
```
Comment (rogatty) [inline front/main/app/models/discussion/forum.js]: `}).catch(this.handleLoadMoreError);`?

File: front/main/app/utils/curated-content-dirty.js [hunk 12]
```diff
@@ -0,0 +1,10 @@
+/**
+ * This small module is too keep state whether curated content was modified in an editor
+ * this way we can track it and show a prompt to a user when he tries to navigate away
+ * with some changes in the editor
+ * 
+ * @type {{isDirty: *}}
+ */
+export const curatedContentState = {
+	isDirty
```
Comment (rogatty) [inline front/main/app/utils/curated-content-dirty.js]:
```
Uncaught ReferenceError: isDirty is not defined
```

**Pull request closed without merging**